### DISCUSSION TITLE
GraphQLContextFactory.generateContext does not support coroutine.

### DISCUSSION THREAD

**Discussion by LEECHHE** [ExpediaGroup/graphql-kotlin#1051] [Q&A] [answered]
Hi, I'm trying to build spring server with version `4.0.0-alpha13` and it seems the guide is unable to be applied on `GraphQLContextFactory`.
```kotlin
interface GraphQLContextFactory<out Context : GraphQLContext, Request> {
    fun generateContext(request: Request): Context?
}
```
Now `GraphQLContextFactory` does not support `suspend` modifier.  I tried to load user entity from DB on ContextFactory as the guide has referred, but the implementation is different.
```kotlin
@Component
class MyGraphQLContextFactory: GraphQLContextFactory<MyGraphQLContext> {
    override suspend fun generateContext(
        request: ServerHttpRequest,
        response: ServerHttpResponse
    ): MyGraphQLContext = MyGraphQLContext(
        myCustomValue = request.headers.getFirst("MyHeader") ?: "defaultValue"
    )
}
```

Can I know workaround to do it?

**Comment by smyrick** [comment]
The context fetching currently is required before we can execute an operation so we explicitly made it non-suspending when we had setup the new `GraphQLContextFactory`. To resolve your issue you will have to run your asynchronous calls in a launch or await.

@dariuszkuc This brings up a good point though, should we be allowing the context fetching and/or request parsing to happen asynchronously? That would mean that everyone must use suspend even if their context building is straight forward

**Reply by dariuszkuc** [reply]
I think it was just a miss and yes it should be suspending - its fine if the resulting implementation won't have any suspensions as it will allow people to load stuff asynchronously if they need to (everything else - server, request parser and handler are already using suspendable functions).

**Comment by smyrick** [comment, marked as answer]
Sorry for moving to fast @LEECHHE, I can not move a discussion back to issues so I opened a new one here: https://github.com/ExpediaGroup/graphql-kotlin/issues/1052
